### ISSUE TITLE
Doodler segmentation error when both MLP and CRF solution has problems

### ISSUE THREAD

**Issue by rtmccune** [dbuscombe-usgs/dash_doodler#42]
The doodled classes are not all appearing in the segmentation. Some of them which take up a much smaller proportion of the image are overwriting larger classes.

I received this error:
```
Traceback (most recent call last):
  File "/opt/miniconda3/envs/dashdoodler/lib/python3.10/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/opt/miniconda3/envs/dashdoodler/lib/python3.10/site-packages/flask/app.py", line 1519, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/opt/miniconda3/envs/dashdoodler/lib/python3.10/site-packages/flask/app.py", line 1517, in full_dispatch_request
    rv = self.dispatch_request()
  File "/opt/miniconda3/envs/dashdoodler/lib/python3.10/site-packages/flask/app.py", line 1503, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/opt/miniconda3/envs/dashdoodler/lib/python3.10/site-packages/dash/dash.py", line 1261, in dispatch
    ctx.run(
  File "/opt/miniconda3/envs/dashdoodler/lib/python3.10/site-packages/dash/_callback.py", line 440, in add_context
    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%
  File "/Users/rtmccune/Desktop/dash_doodler/app.py", line 726, in update_output
    segimgpng, seg, img, color_doodles, doodles  = show_segmentation(
  File "/opt/miniconda3/envs/dashdoodler/lib/python3.10/site-packages/doodler_engine/annotations_to_segmentations.py", line 78, in show_segmentation
    segimg, seg, img, color_doodles, doodles = compute_segmentations(
  File "/opt/miniconda3/envs/dashdoodler/lib/python3.10/site-packages/doodler_engine/annotations_to_segmentations.py", line 415, in compute_segmentations
    color_seg = label_to_colors(seg, img[:,:,0]==0, alpha=128, do_alpha=True, **label_to_colors_args)
  File "/opt/miniconda3/envs/dashdoodler/lib/python3.10/site-packages/doodler_engine/annotations_to_segmentations.py", line 326, in label_to_colors
    cimg[img == c] = colormap[(c + color_class_offset) % len(colormap)]
IndexError: boolean index did not match indexed array along dimension 2; dimension is 3 but corresponding boolean dimension is 4
```
I have included the yml (as txt file for upload to Git), classes.txt, and images that led to this issue. 
![CAM_BF_01_20220515033621_rtm](https://user-images.githubusercontent.com/108094076/197559036-c5e0b441-acef-4396-a3af-3712b696fd44.jpg)
![CAM_BF_01_20220515024222_rtm](https://user-images.githubusercontent.com/108094076/197559053-7a4e23fa-06ce-4d0c-b053-bd0e55558c9b.jpg)
![CAM_BF_01_20220515033621_rtm_overlay](https://user-images.githubusercontent.com/108094076/197559073-54ed706a-d9c6-434c-93ba-bdcca08a9261.png)
![CAM_BF_01_20220515024222_rtm_overlay](https://user-images.githubusercontent.com/108094076/197559077-3f30b6ea-3194-4d27-96b7-d9da3515c3d4.png)

[rtm_dashdoodler.txt](https://github.com/Doodleverse/dash_doodler/files/9852732/rtm_dashdoodler.txt)

[classes.txt](https://github.com/Doodleverse/dash_doodler/files/9852723/classes.txt)

**Comment by ebgoldstein** [comment]
strangely, when i run `gen_images*` on this output, the doodle folder is filled with images that do not actually show doodles.. so its hard to visualize that these are incorrect, but they are..

**Comment by ebgoldstein** [comment]
I think that is the same error @venuswku has seen.. and the same error that @edlazarus has seen

All have seen it with different images.. (and this issue describes it with a different (longer) class.txt file

**Comment by ebgoldstein** [comment]
here are the relevant lines of code from `doodler_engine/annotations_to_segmentations.py`:

https://github.com/Doodleverse/doodler_engine/blob/368db0c53f7719b98888df425a95cf5a60dafbd5/doodler_engine/annotations_to_segmentations.py#L297-L336

```
def label_to_colors(
    img,
    mask,
    alpha,
    colormap,
    color_class_offset,
    do_alpha,
):
    """
    Take MxN matrix containing integers representing labels and return an MxNx4
    matrix where each label has been replaced by a color looked up in colormap.
    colormap entries must be strings like plotly.express style colormaps.
    alpha is the value of the 4th channel
    color_class_offset allows adding a value to the color class index to force
    use of a particular range of colors in the colormap. This is useful for
    example if 0 means 'no class' but we want the color of class 1 to be
    colormap[0].
    """


    colormap = [
        tuple([fromhex(h[s : s + 2]) for s in range(0, len(h), 2)])
        for h in [c.replace("#", "") for c in colormap]
    ]


    cimg = np.zeros(img.shape[:2] + (3,), dtype="uint8")
    minc = np.min(img)
    maxc = np.max(img)


    for c in range(minc, maxc + 1):
        cimg[img == c] = colormap[(c + color_class_offset) % len(colormap)]


    cimg[mask==1] = (0,0,0)


    if do_alpha is True:
        return np.concatenate(
            (cimg, alpha * np.ones(img.shape[:2] + (1,), dtype="uint8")), axis=2
        )
    else:
        return cimg


```

**Comment by ebgoldstein** [comment]
ALSO: 
right before the traceback there is this line:

```
MLP ...
CRF ...
Problem with CRF solution.... reverting back to MLP solution
Exception on /_dash-update-component [POST]
```

**Comment by ebgoldstein** [comment]
This exception is thrown here:
```
        result, unique_labels = do_classify(img,mask,n_sigmas,multichannel,intensity,edges,texture, sigma_min,sigma_max, rf_downsample_value)
        n=len(unique_labels)
        result = result.reshape(img.shape[0],img.shape[1],len(unique_labels))
        # print(result.shape)
        match = np.unique(np.argmax(result,-1))


        logging.info('MLP model applied with sigma range %f : %f' % (sigma_min,sigma_max))
        logging.info('percent RAM usage: %f' % (psutil.virtual_memory()[2]))


        print('CRF ...')


        result2, n = crf_refine(result, img, n,crf_theta_slider_value, crf_mu_slider_value, crf_downsample_factor)#, gt_prob)


        match2 = np.unique(result2-1)
        # print(match2)
        if not np.all(np.array(match)==np.array(match2)):
            print("Problem with CRF solution.... reverting back to MLP solution")
            result2 = result.copy()
```

https://github.com/Doodleverse/doodler_engine/blob/7d0a5aa3a97b19f35b071478fb7435c4c6362b29/doodler_engine/image_segmentation.py#L487-L505

**Comment by ddbuscombe** [comment]
Thanks @rtmccune for raising this issue. I'm looking into troubleshooting. This issue occurred because 

a) the CRF step failed (not yet clear why)
and
b) the MLP step also failed (also not yet clear why)

that - I think - caused the plotting function  `cimg[img == c] = colormap[(c + color_class_offset) % len(colormap)]` to fail


@ebgoldstein you are correct that 

```
        match2 = np.unique(result2-1)
        # print(match2)
        if not np.all(np.array(match)==np.array(match2)):
            print("Problem with CRF solution.... reverting back to MLP solution")
            result2 = result.copy()
```

is probably not a good way to check the validity of the CRF result. what this basically says is that if the unique integer values in the interim MLP result are not the same as the CRF result, then the CRF result gets discarded in favour of the MLP one

However, when the MLP result is invalid, such as apparently here, we have a bigger problem ...

Because the doodler.py and app.py scripts are hard to debug (they contain a mix of GUI functions and codes from `doodler_engine`), I have a troubleshooting script in the `utils` folder called `plot_label_generation.py`. However I discovered that is seriously out of date (it basically hasnt been updated since we moved all the segmentation codes to the `doodler-engine` repo). So first I am troubleshooting that and will hopefully have a new version by end of day. That will help me play with alternative ways to check the validity of the MLP and CRF solution, and to also test the subsequent codes that generated the error

Once that is done, I will look at incorporating the changes back into doodler-engine. I'll update with progress. sorry this one isn't a super fast fix

**Comment by ebgoldstein** [comment]
nice - thx @dbuscombe-usgs -  :muscle:

**Comment by ddbuscombe** [comment]
Update. I rewrote the `utils/plot_label_generation.py` script so it uses the same codes as ['doodler-engine\](https://github.com/Doodleverse/doodler_engine/blob/main/doodler_engine/image_segmentation.py) to carry out the MLP and CRF segmentations

It produces the following outputs for the offending image, which all look fine

Standardized image:
![CAM_BF_01_20220515024222_rtm_image_filt_labelgen](https://user-images.githubusercontent.com/3596509/197646670-0a768038-ca8f-40d9-a28f-7346a8f98b3f.png)

Image features used by MLP
![CAM_BF_01_20220515024222_rtm_image_feats_labelgen](https://user-images.githubusercontent.com/3596509/197646669-01f2d4e2-c97d-42ec-a246-8cee6eb54ee6.png)

MLP output:
![CAM_BF_01_20220515024222_rtm_image_label_MLP_labelgen](https://user-images.githubusercontent.com/3596509/197646683-fed90068-56bd-418d-8931-90a22cf882a9.png)

CRF output:
![CAM_BF_01_20220515024222_rtm_image_label_CRF_labelgen](https://user-images.githubusercontent.com/3596509/197646677-2f626fa8-d876-4e50-a1df-25133084cdef.png)

Greyscale label image (correct that it is dark)
![CAM_BF_01_20220515024222_rtm_label_labelgen](https://user-images.githubusercontent.com/3596509/197646687-3bffb3fa-5d6b-4edc-989e-e4f7ee177fd1.png)

Doodles (they are very thin so invisible - but I verified they are there)
![CAM_BF_01_20220515024222_rtm_image_doodles_labelgen](https://user-images.githubusercontent.com/3596509/197646664-4df4d992-a424-43bf-9aa0-6cc9b47a9dce.png)

Doodles greyscale image
![CAM_BF_01_20220515024222_rtm_doodles_labelgen](https://user-images.githubusercontent.com/3596509/197646663-f9cfda0d-2f54-4548-8664-310137985021.png)

I will wait to push these changes until I've verified that no modifications need to be made to the `app.py` function

Next, I will address the issue identified above, in that there is currently no fallback if the MLP fails. Therefore the following code is creating a problem:

https://github.com/Doodleverse/doodler_engine/blob/main/doodler_engine/image_segmentation.py#L500-L504

The proposed solution is to, instead of reverting back to an erroneous MLP solution, to compute the CRF using different inputs. Currently, the MLP creates softmax scores that the CRF uses to make integer labels. However, it is also possible to use the doodles directly with the CRF algorithm, to bypass the MLP when it fails. Fingers crossed that fixes it! 🤞

**Comment by ddbuscombe** [comment]
Okay, I see further issues. As @rtmccune noted, the doodled classes are not appearing. Not sure how this has gone unnoticed for this long (perhaps because we generally use lists of classes that are always present?)

Classes 2,4,3,6  (sidewalk, vehicle, person, other) were doodled, but the segmentation was returning classes 1,2,3,4 - completely different! 

The proposed fix is to make the MLP result softmax scores into a label (argmax), then find the unique integers in the original doodled image and the MLP label, then iterate through them to recode

```
            mlp_result = np.argmax(mlp_result,-1)+1

            uniq_doodles = np.unique(data['doodles'])[1:]
            uniq_mlp = np.unique(mlp_result)
            mlp_result2 = np.zeros_like(mlp_result)
            for o,e in zip(uniq_doodles,uniq_mlp):
                mlp_result2[mlp_result==e] = o

            mlp_result = mlp_result2.copy()
```

The same would be done for the CRF output... more soon

**Comment by ddbuscombe** [comment]
To pass the MLP properly to CRF, the MLP is one-hot encoded like this

```
nx, ny = mlp_result.shape
mlp_result_softmax = np.zeros((nx,ny,NUM_LABEL_CLASSES))
mlp_result_softmax[:,:,:NUM_LABEL_CLASSES] = (np.arange(NUM_LABEL_CLASSES) == 1+mlp_result[...,None]-1).astype(int)
```
which can be checked for validity using the below comparison, i.e. are the unique values in the one-hot encoded array the same as the doodles?

```
if np.all(uniq_doodles==np.unique(np.argmax(mlp_result_softmax,-1))):
    print("that worked")
```

then CRF from softmax scores:

```
crf_result, n = crf_refine(mlp_result_softmax, data['orig_image'], 
                        NUM_LABEL_CLASSES,DEFAULT_CRF_THETA, 
                        DEFAULT_CRF_MU, DEFAULT_CRF_DOWNSAMPLE)
```

followed by the same kludge as earlier to remap classes

```
uniq_crf = np.unique(crf_result)
crf_result2 = np.zeros_like(crf_result)
for o,e in zip(uniq_doodles,uniq_crf):
    crf_result2[crf_result==e] = o

crf_result = crf_result2.copy()
```

This results in near-identical MLP and CRF outputs:

![CAM_BF_01_20220515024222_rtm_image_label_MLP_labelgen](https://user-images.githubusercontent.com/3596509/197679742-f2eed787-2968-4b87-a727-205932c8ea0b.png)

![CAM_BF_01_20220515024222_rtm_image_label_CRF_labelgen](https://user-images.githubusercontent.com/3596509/197679733-c8c908d1-ac4c-4f2f-8f8d-f828b0830ff9.png)

I need to test further tomorrow . Meanwhile @ebgoldstein lmk if you see any holes in this logic?

I also made a new function

```
crf_result, n = crf_refine_from_integer_labels(data['doodles'], data['orig_image'], NUM_LABEL_CLASSES,
                                            DEFAULT_CRF_THETA, DEFAULT_CRF_MU, 
                                            DEFAULT_CRF_DOWNSAMPLE)
 ```

this is much simpler because it calls the doodles directly. It also needs the remapping kludge...

```
uniq_crf = np.unique(crf_result)
crf_result2 = np.zeros_like(crf_result)
for o,e in zip(uniq_doodles,uniq_crf):
    crf_result2[crf_result==e] = o

crf_result = crf_result2.copy()
```

However, the results are not good:

![CAM_BF_01_20220515024222_rtm_image_label_CRF_labelgen](https://user-images.githubusercontent.com/3596509/197680276-0ac41997-03bf-4153-bcb4-3a953564c76f.png)

which is why we discarded this approach before .... that makes sense to me - mush better to pass the softmax scores than the argmax operation to CRF, therefore I propose we only revert to that solution as a last resort, i.e.

```
if not np.all(uniq_doodles==np.unique(crf_result)):
    print("that did not work!")

    crf_result, n = crf_refine_from_integer_labels(data['doodles'], data['orig_image'], NUM_LABEL_CLASSES,
                                                DEFAULT_CRF_THETA, DEFAULT_CRF_MU, 
                                                DEFAULT_CRF_DOWNSAMPLE)

    uniq_crf = np.unique(crf_result)
    crf_result2 = np.zeros_like(crf_result)
    for o,e in zip(uniq_doodles,uniq_crf):
        crf_result2[crf_result==e] = o

    crf_result = crf_result2.copy()
```

For reference, the new `crf_refine_from_integer_labels` function is identical to the softmax-based one, except the unary potentials are defined from labels (`from pydensecrf.utils import unary_from_labels`)


```
##========================================================
def crf_refine_from_integer_labels(label,
    img,n,
    crf_theta_slider_value,
    crf_mu_slider_value,
    crf_downsample_factor): #gt_prob
    """
    "crf_refine(label, img)"
    This function refines a label image based on an input label image and the associated image
    Uses a conditional random field algorithm using spatial and image features
    INPUTS:
        * label [ndarray]: label image 2D matrix of integers
        * image [ndarray]: image 3D matrix of integers
    OPTIONAL INPUTS: None
    GLOBAL INPUTS: None
    OUTPUTS: label [ndarray]: label image 2D matrix of integers
    """

    Horig = img.shape[0]
    Worig = img.shape[1]
    l_unique = len(np.unique(label)) #label.shape[-1]

    # label = label.reshape(Horig,Worig,l_unique)

    scale = 1+(5 * (np.array(img.shape).max() / 3000))
    logging.info('CRF scale: %f' % (scale))

    logging.info('CRF downsample factor: %f' % (crf_downsample_factor))
    logging.info('CRF theta parameter: %f' % (crf_theta_slider_value))
    logging.info('CRF mu parameter: %f' % (crf_mu_slider_value))

    # decimate by factor by taking only every other row and column
    img = img[::crf_downsample_factor,::crf_downsample_factor, :]
    # do the same for the label image
    label = label[::crf_downsample_factor,::crf_downsample_factor]
    # yes, I know this aliases, but considering the task, it is ok; the objective is to
    # make fast inference and resize the output

    logging.info('Images downsampled by a factor of %f' % (crf_downsample_factor))

    H = img.shape[0]
    W = img.shape[1]
    U = unary_from_labels(label.astype('int'), n, gt_prob=0.51) #np.argmax(label,-1)
    d = dcrf.DenseCRF2D(H, W, n)

    # U = unary_from_softmax(np.ascontiguousarray(np.rollaxis(label,-1,0)))
    # d = dcrf.DenseCRF2D(H, W, l_unique)

    d.setUnaryEnergy(U)

    # to add the color-independent term, where features are the locations only:
    d.addPairwiseGaussian(sxy=(3, 3),
                 compat=3,
                 kernel=dcrf.DIAG_KERNEL,
                 normalization=dcrf.NORMALIZE_SYMMETRIC)
    feats = create_pairwise_bilateral(
                          sdims=(crf_theta_slider_value, crf_theta_slider_value),
                          schan=(scale,scale,scale),
                          img=img,
                          chdim=2)

    d.addPairwiseEnergy(feats, compat=crf_mu_slider_value, kernel=dcrf.DIAG_KERNEL,normalization=dcrf.NORMALIZE_SYMMETRIC) #260

    logging.info('CRF feature extraction complete ... inference starting')

    Q = d.inference(10)
    result = np.argmax(Q, axis=0).reshape((H, W)).astype(np.uint8) +1
    logging.info('CRF inference made')

    # uniq = np.unique(result.flatten())

    result = resize(result, (Horig, Worig), order=0, anti_aliasing=False) #True)
    result = rescale(result, 1, l_unique).astype(np.uint8)

    # result = rescale(result, orig_mn, orig_mx).astype(np.uint8)

    logging.info('label resized and rescaled ... CRF post-processing complete')

    return result, l_unique
```

However, tomorrow I will play with more iterations, e.g. `Q = d.inference(30)` and perhaps a different `gt_prob` (probability prior)

**Comment by ddbuscombe** [comment]
I tested on all 3 provided images and results look reasonable

![CAM_BF_01_20220515033621_rtm_image_label_CRF_labelgen](https://user-images.githubusercontent.com/3596509/197682578-c254780c-2dfc-49e0-b0dc-7ccc5ff153f2.png)

![CAM_BF_01_20220515141822_rtm_image_label_CRF_labelgen](https://user-images.githubusercontent.com/3596509/197682585-67e667e4-548f-4746-b6fd-324b8fae43aa.png)

![CAM_BF_01_20220515024222_rtm_image_label_CRF_labelgen](https://user-images.githubusercontent.com/3596509/197682571-372cb0de-2703-415a-896d-f4936165c88d.png)


If this is a go, next step will be to update the `segmentation` function inside `doodler_engine/image_segmentation.py`

This would need to be thoroughly tested before finalized as a new release on pypi

**Comment by ebgoldstein** [comment]
this makes sense for sure. I can definitely spend a few hours testing later today. 

The Qs in my mind are mostly about failure of the MLP-> CRF route, and this statement: `np.all(uniq_doodles==np.unique(crf_result))`
- when would this be false?
- why would this be false?
- If this is false (and an operator is presented with CRF output from integer labels), what can they do to get a better segmentation (i.e., get the statement to be true and get back to the softmax-route) ? Doodle more? Delete doodles?

**Comment by ddbuscombe** [comment]
* that statement would be false if the doodled classes did not match the CRF classes, such as when the labeled classes are non-monotonic
* that would occur if the CRF solution failed to converge so badly that it did not predict the full range of classes from the softmax scores, for example if it determined that one or more classes doodled were not sufficiently consistent. I build in a fallback; instead if using the softmax scores, it would use the doodles directly. 
* There could be other reasons why it would occur, in the form of software bugs 😆 
* Yes, delete doodles, doodle more, change parameters such as downsample .... other things could be possible with a more major software overhaul, such as a different classifier (other than MLP)

I have made some further changes which now implements (in `def segmentation()` in `doodler_engine\image_segmentation.py`) the following algorithm:
* do MLP
* if number of classes in MLP is not the same as the original doodles, do CRF from original doodles (if the above fails, return the MLP result. Note this never occurred in my trials - the CRF from original doodles worked every time)
* if the number of classes in the MLP is the same as the the original doodles, do CRF from MLP softmax scores

I tested with some large and small images, and with a class list of 5 classes. I labeled at least the following class combinations 

- 1,2,3,4, and 5
- 1 and 5
- 1 and 2
- 2, 4, and 5
- 2 and 5

In each case, the new implementation produced good outputs with no errors (see below). I've also updated `utils/plot_label_generation.py` script (which is what I used to work out the next formula)

I went ahead and implemented the changes in doodler-engine(https://github.com/Doodleverse/doodler_engine), and in a dev branch of doodler (https://github.com/Doodleverse/dash_doodler/tree/bugfix)

To test:
1. clone the `bugfix` branch of dash_doodler
2. clone doodler_engine
3. cd to doodler_engine repo, and activate `dashdoodler` conda env
4. `pip install -e .`
5. cd to dash_doodler repo
6. use doodler.py to generate npz files
7. launch `utils\plot_label_generation.py` and point to the new npz files
8. launch `utils\plot_label_generation.py` and point to any old npz files you may have

![2018-09-PelicanIsland-ortho-rgb-5cm_09_07_ID3_Enter-user-ID_image_label_CRF_labelgen](https://user-images.githubusercontent.com/3596509/198419810-cd307d4a-3b51-4ba8-86c6-a266fde69d56.png)
![20181224T154641_20181224T155116_T18SVF_lrr_ID1_Enter-user-ID_image_label_CRF_labelgen](https://user-images.githubusercontent.com/3596509/198419814-7da975b3-8db4-4abc-8bd0-cfb665404fce.png)
![C20056232-3_Enter-user-ID_image_label_CRF_labelgen](https://user-images.githubusercontent.com/3596509/198419817-67a152fd-564d-4819-b5fd-7fa79ad1f506.png)
![problem_image_Enter-user-ID_image_label_CRF_labelgen](https://user-images.githubusercontent.com/3596509/198419818-8a479a96-98a6-4e79-87d3-e7d1c145697e.png)

**Comment by ebgoldstein** [comment]
4 classes in class.txt (water, sand, veg, dev)
this image has only water class (image below terminal message).

terminal output:
```
/home/ebgoldstein/anaconda3/envs/dashdoodlerNew/lib/python3.8/site-packages/skimage/io/manage_plugins.py:23: UserWarning: Your installed pillow version is < 8.1.2. Several security issues (CVE-2021-27921, CVE-2021-25290, CVE-2021-25291, CVE-2021-25293, and more) have been fixed in pillow 8.1.2 or higher. We recommend to upgrade this library.
  from .collection import imread_collection_wrapper
Hyperparameters imported from my_defaults.py
Go to http://127.0.0.1:8050/ in your web browser to use Doodler
Dash is running on http://127.0.0.1:8050/

 * Serving Flask app 'app'
 * Debug mode: off
Exception on /_dash-update-component [POST]
Traceback (most recent call last):
  File "/home/ebgoldstein/anaconda3/envs/dashdoodlerNew/lib/python3.8/site-packages/flask/app.py", line 2525, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/ebgoldstein/anaconda3/envs/dashdoodlerNew/lib/python3.8/site-packages/flask/app.py", line 1822, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/ebgoldstein/anaconda3/envs/dashdoodlerNew/lib/python3.8/site-packages/flask/app.py", line 1820, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/ebgoldstein/anaconda3/envs/dashdoodlerNew/lib/python3.8/site-packages/flask/app.py", line 1796, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/ebgoldstein/anaconda3/envs/dashdoodlerNew/lib/python3.8/site-packages/dash/dash.py", line 1261, in dispatch
    ctx.run(
  File "/home/ebgoldstein/anaconda3/envs/dashdoodlerNew/lib/python3.8/site-packages/dash/_callback.py", line 440, in add_context
    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%
  File "/home/ebgoldstein/Documents/ResearchIdeas/CHIIP/dash_doodler-bugfix/app.py", line 726, in update_output
    segimgpng, seg, img, color_doodles, doodles  = show_segmentation(
  File "/home/ebgoldstein/Documents/ResearchIdeas/CHIIP/doodler_engine-main/doodler_engine/annotations_to_segmentations.py", line 78, in show_segmentation
    segimg, seg, img, color_doodles, doodles = compute_segmentations(
  File "/home/ebgoldstein/Documents/ResearchIdeas/CHIIP/doodler_engine-main/doodler_engine/annotations_to_segmentations.py", line 401, in compute_segmentations
    seg = segmentation(img, mask,#img_path, results_folder, callback_context,
  File "/home/ebgoldstein/Documents/ResearchIdeas/CHIIP/doodler_engine-main/doodler_engine/image_segmentation.py", line 700, in segmentation
    return crf_result
UnboundLocalError: local variable 'crf_result' referenced before assignment
```
![S28474284-3](https://user-images.githubusercontent.com/5330599/198672111-85491b9b-f522-4dfb-b50e-f991be942b96.jpg)

**Comment by ebgoldstein** [comment]
i have confirmed that its not just for 'labeled as all class 1' images.. it also fails with same error message for 'labeled as all class 3' images... so it is failing when only a single label is provided

**Comment by ddbuscombe** [comment]
Thanks. I have fixed that bug now (thankfully, very trivial). Please try with the latest https://github.com/Doodleverse/doodler_engine

**Comment by ebgoldstein** [comment]
tested - now works as expected - :tada:

**Comment by ddbuscombe** [comment]
After a bunch more testing, I can also confirm that I'm not ever seeing the original error, and also that the results are now correct under a range of labeling scenarios and test imagery

Once the PR is merged, and a new doodler-engine is posted to pypi, I will let you know. Then, I'll keep this issue open in case @rtmccune wants to do further tests

**Comment by ddbuscombe** [comment]
Okay, PR is merged. @rtmccune please test this new version by first doing a `git pull` (or a fresh clone), then in your activated `dashdoodler` environment, `pip install doodler-engine -U` (the `-U` stands for 'upgrade' and should install the latest version v0.0.3) 

thanks!

**Comment by ebgoldstein** [comment]
I feel good about closing this for now.. :crossed_fingers: 
great work everyone!

**Comment by ebgoldstein** [comment]
@dbuscombe-usgs - here is something from @rtmccune. He doodled ~ 20 or so images. most came out fine, but 2 ended up looking strange (wrong classes). here is an example of the failure :

Note 
1) that the MLP solution was good, the CRF was bad.
2) the final solution looked like the CRF solution, but switched the class (from class 1 (red) to class 0 (blue))

Image:
![S21766540-1_L12](https://user-images.githubusercontent.com/5330599/199042541-ca41ffdd-6314-4ad4-92e1-1daa9febf3a8.jpg)

Doodles:
![S21766540-1_L12_doodles](https://user-images.githubusercontent.com/5330599/199042555-0e6134eb-f372-4765-8bdb-23385c3fec8c.png)

MLP:
![S21766540-1_L12_image_label_MLP_labelgen](https://user-images.githubusercontent.com/5330599/199042802-e4ba22ac-1f49-483a-a41f-bbad53520d4c.png)

CRF
![S21766540-1_L12_image_label_CRF_labelgen](https://user-images.githubusercontent.com/5330599/199042818-f9a28c28-9664-4b60-aa2b-75033c89d21d.png)

Overlay:
![S21766540-1_L12_overlay](https://user-images.githubusercontent.com/5330599/199042581-4a63827b-50d3-48a8-a3e9-9495d7de2ea8.png)

**Comment by ddbuscombe** [comment]
Looks like the MLP gets a good result and the CRF does not, but the program defaults to saving the CRF output

I dont have any quick/easy suggestions for fixing it, but `utils\https://github.com/Doodleverse/dash_doodler/blob/main/utils/plot_label_generation.py` can regenerate both the CRF and MLP labels, and you can decide which one you like best

Perhaps going forward, Doodler creates both MLP and CRF outputs?

**Comment by ebgoldstein** [comment]
i am going to close this.. I made a new issue #46 to discuss falling back to MLP outputs
